librz/asm/p/asm_java.c

```diff
@@ -8,40 +7,131 @@
 
-//#include "../../bin/format/java/class.h"
-#include "../arch/java/code.h"
-
-static int disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
-	RzBinJavaObj *obj = NULL;
-	RzBin *bin = a->binb.bin;
-	RzBinPlugin *plugin = bin && bin->cur && bin->cur->o ? bin->cur->o->plugin : NULL;
-	if (plugin && plugin->name) {
-		if (!strcmp(plugin->name, "java")) { // XXX slow
-			obj = bin->cur->o->bin_obj; //o;
-			//eprintf("Handling: %s disasm.\n", b->cur.file);
+#include "../arch/java/jvm.h"
+
+typedef struct java_analysis_context_t {
+	LookupSwitch ls;
+	TableSwitch ts;
+	ut16 switchop;
+	ut64 pc;
+	ut64 last;
+	ut32 count;
+} JavaAsmContext;
+
+static void update_context(JavaAsmContext *ctx) {
+	ctx->count++;
+	if (ctx->switchop == BYTECODE_AA_TABLESWITCH && ctx->count > ctx->ts.length) {
+		ctx->switchop = BYTECODE_00_NOP;
+	} else if (ctx->switchop == BYTECODE_AB_LOOKUPSWITCH && ctx->count > ctx->ls.npairs) {
+		ctx->switchop = BYTECODE_00_NOP;
+	}
+}
+
+static int java_disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
+	JavaAsmContext *ctx = (JavaAsmContext *)a->plugin_data;
+	rz_strbuf_set(&op->buf_asm, "invalid");
+
+	if (a->pc < ctx->last) {
+		ctx->switchop = BYTECODE_00_NOP;
+	}
+	ctx->last = a->pc;
+	switch (ctx->switchop) {
+	case BYTECODE_AA_TABLESWITCH: {
+		if (len < 4) {
+			eprintf("[!] java_analysis: no enough data for lookupswitch case.\n");
+			return -1;
+		}
+		op->size = 4;
+		ut64 jump = ctx->pc + rz_read_be32(buf);
+		rz_strbuf_setf(&op->buf_asm, "case %d: goto 0x%" PFMT64x, ctx->count + ctx->ts.low, jump);
+		update_context(ctx);
+		return op->size;
+	}
+	case BYTECODE_AB_LOOKUPSWITCH: {
+		if (len < 8) {
+			eprintf("[!] java_analysis: no enough data for lookupswitch case.\n");
+			return -1;
 		}
+		op->size = 8;
+		st32 number = (st32)rz_read_be32(buf);
+		ut64 jump = ctx->pc + rz_read_at_be32(buf, 4);
+		rz_strbuf_setf(&op->buf_asm, "case %d: goto 0x%" PFMT64x, number, jump);
+		update_context(ctx);
+		return op->size;
+	}
+	default:
+		break;
+	}
+
+	JavaVM vm = { 0 };
+	Bytecode bc = { 0 };
+
+	rz_strbuf_set(&op->buf_asm, "invalid");
+
+	ut64 section = a->pc;
+	if (a->binb.bin) {
+		const RzBinSection *sec = a->binb.get_vsect_at(a->binb.bin, a->pc);
+		if (sec) {
+			section = sec->paddr;
+		}
+	}
+
+	if (!jvm_init(&vm, buf, len, a->pc, section)) {
+		eprintf("[!] java_disassemble: bad or invalid data.\n");
+		return -1;
+	}
+	op->size = 1;
+	if (jvm_fetch(&vm, &bc)) {
+		op->size = bc.size;
+		bytecode_snprint(&op->buf_asm, &bc);
+		if (bc.opcode == BYTECODE_AA_TABLESWITCH) {
+			ctx->count = 0;
+			ctx->switchop = BYTECODE_AA_TABLESWITCH;
+			ctx->ts = *((TableSwitch *)bc.extra);
+			ctx->pc = a->pc;
+		} else if (bc.opcode == BYTECODE_AB_LOOKUPSWITCH) {
+			ctx->count = 0;
+			ctx->switchop = BYTECODE_AB_LOOKUPSWITCH;
+			ctx->ls = *((LookupSwitch *)bc.extra);
+			ctx->pc = a->pc;
+		}
+		bytecode_clean(&bc);
+	} else {
+		eprintf("[!] java_disassemble: jvm fetch failed.\n");
```

Comment:
I would use RZ_LOG_ERROR instead, so you can filter the output or redirect to file, etc.

Reply:
In general in the plugins, I mean.